src/ray/core_worker/object_interface.cc

```diff
@@ -1,25 +1,142 @@
+
+  auto &context = core_worker_.GetContext();
+  bool was_blocked = false;
+
+  std::unordered_map<ObjectID, int> unready;
+  for (int i = 0; i < ids.size(); i++) {
+    unready.insert({ids[i], i});
+  }
+
+  int num_attempts = 0;
+  bool should_break = false;
+  int64_t remaining_timeout = timeout_ms;
+  // Repeat until we get all objects.
+  while (!unready.empty() && !should_break) {
+    std::vector<ObjectID> unready_ids;
+    for (const auto &entry : unready) {
+      unready_ids.push_back(entry.first);
+    }
+
+    // For the initial fetch, we only fetch the objects, do not reconstruct them.
+    bool fetch_only = num_attempts == 0;
+    if (!fetch_only) {
+      // If fetch_only is false, this worker will be blocked.
+      was_blocked = true;
+    }
+
+    // TODO: can call `fetchOrReconstruct` in batches as an optimization.
+    RAY_CHECK_OK(core_worker_.raylet_client_->FetchOrReconstruct(
+        unready_ids, fetch_only, context.GetCurrentTaskID()));
+
+    // Get the objects from the object store, and parse the result.
+    int64_t get_timeout = (remaining_timeout < 0 || remaining_timeout > c_get_timeout_ms_) ?
+        c_get_timeout_ms_ : remaining_timeout;
+    should_break = (get_timeout == remaining_timeout);
+  
+    if (remaining_timeout >= c_get_timeout_ms_) {
+      remaining_timeout -= c_get_timeout_ms_;
+    }
+
+    std::vector<plasma::ObjectID> plasma_ids;
+    for (const auto &id : unready_ids) {
+      plasma_ids.push_back(id.to_plasma_id());
+    }
+
+    std::vector<plasma::ObjectBuffer> object_buffers;
+    auto status = store_client_.Get(plasma_ids, get_timeout, &object_buffers);
+  
+    for (int i = 0; i < object_buffers.size(); i++) {
+      if (object_buffers[i].data != nullptr) {
+        const auto &object_id = unready_ids[i];
+        (*results)[unready[object_id]] = std::make_shared<PlasmaBuffer>(object_buffers[i].data);
+        unready.erase(object_id);
+      }
+    }
+
+    num_attempts += 1;
+    // TODO: log a message if attempted too many times.
+  }
+
+  if (was_blocked) {
+    RAY_CHECK_OK(core_worker_.raylet_client_->NotifyUnblocked(context.GetCurrentTaskID()));
+  }
+
   return Status::OK();
 }
 
 Status CoreWorkerObjectInterface::Wait(const std::vector<ObjectID> &object_ids,
                                        int num_objects, int64_t timeout_ms,
                                        std::vector<bool> *results) {
-  return Status::OK();
+  auto &context = core_worker_.GetContext();
+  WaitResultPair result_pair;
+  auto status = core_worker_.raylet_client_->Wait(object_ids, num_objects, timeout_ms,
+      false, context.GetCurrentTaskID(), &result_pair);
+  std::unordered_set<ObjectID> ready_ids;
+  for (const auto &entry : result_pair.first) {
+    ready_ids.insert(entry);
+  }
+
+  (*results).resize(object_ids.size());
+  for (int i = 0; i < object_ids.size(); i++) {
+    (*results)[i] = ready_ids.count(object_ids[i]) > 0;
+  }
```

Comment:
maybe add a TODO here to also change raylet_client.wait to return a bit set?
Otherwise we'll need to convert ID to bitset, and then ID again.